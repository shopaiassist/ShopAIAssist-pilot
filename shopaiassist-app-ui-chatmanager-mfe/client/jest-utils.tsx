import React, { JSXElementConstructor, ReactElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouterProps } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';
import { merge } from 'ts-deepmerge';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { ByRoleMatcher, render, RenderResult } from '@testing-library/react';
import { reducer, RootState, store as baseStore } from './src/store/store';
import i18nInstance from './src/utils/i18n';

type InitialEntries = MemoryRouterProps['initialEntries'];
type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};
type PartialRootState = DeepPartial<RootState>;

interface RenderOptions {
  entries?: InitialEntries;
  state?: PartialRootState;
  store?: EnhancedStore;
}

interface SnapshotOptions {
  isModal?: boolean | ByRoleMatcher;
  state?: PartialRootState;
}

interface WrapperProps {
  children: ReactElement;
}

// export const mockApi = (apiInstance: AxiosInstance) => ({
//   delete: jest.spyOn(apiInstance, 'delete'),
//   get: jest.spyOn(apiInstance, 'get'),
//   post: jest.spyOn(apiInstance, 'post'),
//   put: jest.spyOn(apiInstance, 'put'),
//   patch: jest.spyOn(apiInstance, 'patch'),
// });

/** Redux store with optional default content */
export const mockStore = (state: PartialRootState = {}): EnhancedStore =>
  configureStore({ preloadedState: merge(baseStore.getState(), state), reducer: reducer as never });

/** Render function with access to override store/state */
export const renderAll = (ui: ReactElement, { entries, state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const Wrapper = ({ children }: WrapperProps) => (
    <MemoryRouter initialEntries={entries}>
      <Provider store={data}>
        <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
      </Provider>
    </MemoryRouter>
  );
  return render(ui, { wrapper: Wrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with i18n provider */
export const renderWithI18n = (ui: ReactElement): RenderResult => {
  const I18nWrapper = ({ children }: WrapperProps) => <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>;
  return render(ui, { wrapper: I18nWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with access to override store/state */
export const renderWithRedux = (ui: ReactElement, { state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const ReduxWrapper = ({ children }: WrapperProps) => (
    <Provider store={data}>
      <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
    </Provider>
  );
  return render(ui, { wrapper: ReduxWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Get content to test against a snapshot */
export const getSnapshot = (ui: ReactElement, state?: PartialRootState): ChildNode =>
  renderAll(ui, { state }).container.firstChild as ChildNode;

/** Get modal content to test against a snapshot */
export const getSnapshotModal = (ui: ReactElement, state?: PartialRootState, role?: ByRoleMatcher): ChildNode =>
  renderAll(ui, { state }).getByRole(role || 'presentation');

/** Helper function to generate snapshots */
export const snapshot = (ui: ReactElement, { isModal, state }: SnapshotOptions = {}) => {
  const role = typeof isModal === 'boolean' ? undefined : isModal;
  // eslint-disable-next-line no-extra-boolean-cast
  const snapshot = !!isModal ? getSnapshotModal(ui, state, role) : getSnapshot(ui, state);
  expect(snapshot).toMatchSnapshot();
};

// re-export everything
export * from '@testing-library/react';
