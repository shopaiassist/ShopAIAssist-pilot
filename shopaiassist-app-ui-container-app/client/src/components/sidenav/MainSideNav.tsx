import React from 'react';
import { i18n } from '@/';
import { SafSideNav } from '@/core-components/react';
import classNames from 'classnames';
import { useNavigate } from 'react-router-dom';

import { OLYMPUS_ROUTES } from '../../App';
import useMattersSideNav from '../../hooks/useMattersSideNav';
import useMFECommunication from '../../hooks/useMFECommunication';
import useSelectedMenuItem from '../../hooks/useSelectedMenuItem';
import MainFooter from '../footer/MainFooter';
import NavMenuItemComponent from './NavMenuItem';

import './SideNav.scss';

interface MainSideNavProps {
  expanded: boolean;
  isMobile: boolean;
  toggleExpanded: (expanded?: boolean) => void;
  setSecondaryNavState: (state: SecondaryNavState) => void;
}

const MainSideNav = ({ expanded, isMobile, toggleExpanded, setSecondaryNavState }: MainSideNavProps) => {
  const [sendEvent] = useMFECommunication('isSortingMenu');
  const navigate = useNavigate();
  const { t } = i18n.useTranslation();

  const [, , handleMyWorkClick] = useMattersSideNav();

  /** List of nav menu items to render */
  const navMenuItems: NavMenuTree = {
    topGroup: [
      {
        id: 'my-work-nav-item',
        icon: 'sparkles',
        bold: true,
        label: t('CHAT_HISTORY'),
        showArrow: true,
        navPath: OLYMPUS_ROUTES.WORK,
        onClick: handleMyWorkClick,
        pendoId: 'chat-history-button',
      },
      // {
      //   id: 'matters-nav-item',
      //   icon: 'database',
      //   bold: true,
      //   label: t('DATABASES'),
      //   navPath: OLYMPUS_ROUTES.DATABASES,
      // },
      // {
      //   id: 'manage-nav-item',
      //   icon: 'gear',
      //   bold: true,
      //   label: t('MANAGE_ShopAIAssist'),
      //   showArrow: true,
      //   secondaryNavTree: 'manage',
      // },
      // TODO: Admin settings got pushed to post-release. Re-enable when ready.
      // {
      //   id: 'admin-settings-nav-item',
      //   icon: 'sitemap',
      //   bold: true,
      //   label: t('ADMIN_SETTINGS'),
      //   showArrow: true,
      //   secondaryNavTree: 'admin',
      // },
    ],
    bottomGroup: [
      // {
      //   id: 'whats-new-nav-item',
      //   icon: 'bell',
      //   bold: true,
      //   label: t('WHATS_NEW'),
      //   url: 'https://www.example.com/en/release-notes.html', // TODO: Need final URL for this
      // },
      // {
      //   id: 'support-nav-item',
      //   icon: 'circle-question',
      //   bold: true,
      //   label: t('SUPPORT'),
      //   showArrow: true,
      //   secondaryNavTree: 'support',
      // },
      // {
      //   id: 'legal-nav-item',
      //   icon: 'shield',
      //   bold: true,
      //   label: t('LEGAL_INFORMATION'),
      //   showArrow: true,
      //   secondaryNavTree: 'legal',
      // },
    ],
  };

  const [selectedMenuItem, setSelectedMenuItem] = useSelectedMenuItem(navMenuItems);

  /** Handle selecting menu items */
  const handleMenuItemClick = (menuItem: NavMenuItem) => {
    setSelectedMenuItem(menuItem.id);
    if (menuItem.secondaryNavTree) {
      setSecondaryNavState({ show: true, tree: menuItem.secondaryNavTree });
    } else {
      setSecondaryNavState({ show: false, tree: 'manage' });
      sendEvent({ message: 'isSortingMenu', body: { isSortingMenu: true } });
    }
    if (menuItem.navPath) {
      if (isMobile) toggleExpanded(false);
      navigate(menuItem.navPath);
    }
    menuItem.onClick?.();
  };

  /** Function to render a NavMenuItemComponent */
  const renderNavMenuItem = (menuItem: NavMenuItem) => {
    const { bold, id, icon, label, showArrow, url, pendoId } = menuItem;
    return (
      <NavMenuItemComponent
        id={id}
        key={id}
        onClick={() => handleMenuItemClick(menuItem)}
        icon={icon}
        label={label}
        showArrow={showArrow}
        bold={bold}
        selected={selectedMenuItem === id}
        url={url}
        hasLink={!!url}
        pendoId={pendoId}
      />
    );
  };

  return (
    <SafSideNav
      id="main-side-nav"
      className={classNames('olympus-side-nav main-side-nav', { expanded: expanded })}
      state={expanded ? 'open' : 'closed'}
      openAriaLabel="Expand main navigation"
      closeAriaLabel="Collapse main navigation"
      aria-label="Main navigation"
      openSvgName="arrow-right-from-line"
      closeSvgName="arrow-left-from-line"
      onClose={() => {
        toggleExpanded(false);
      }}
      onOpen={() => {
        toggleExpanded(true);
      }}
    >
      <div className="main-nav-container">
        <div className="nav-group">{navMenuItems.topGroup.map(renderNavMenuItem)}</div>
        <div className="nav-group">
          {navMenuItems.bottomGroup.map(renderNavMenuItem)}
          {expanded && <MainFooter isMobile={isMobile} />}
        </div>
      </div>
    </SafSideNav>
  );
};

export default MainSideNav;
