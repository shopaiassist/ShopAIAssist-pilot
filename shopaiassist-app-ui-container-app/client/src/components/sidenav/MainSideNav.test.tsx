// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { renderAll, snapshot, waitFor } from '../../../jest-utils';
import MainSideNav from './MainSideNav';

describe('MainSideNav', () => {
  const mockToggleExpanded = jest.fn();
  const mockSetSecondaryNavState = jest.fn();

  const props = {
    expanded: true,
    toggleExpanded: mockToggleExpanded,
    setSecondaryNavState: mockSetSecondaryNavState,
  };

  describe('render', () => {
    it('should render', async () => {
      snapshot(<MainSideNav {...props} isMobile={false} />);
    });
  });
});
