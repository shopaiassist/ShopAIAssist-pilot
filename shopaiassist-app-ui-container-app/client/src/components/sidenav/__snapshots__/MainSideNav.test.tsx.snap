// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainSideNav render should render 1`] = `
<div
  aria-label="Main navigation"
  class="olympus-side-nav main-side-nav expanded"
  closearialabel="Collapse main navigation"
  closesvgname="arrow-left-from-line"
  id="main-side-nav"
  openarialabel="Expand main navigation"
  opensvgname="arrow-right-from-line"
  state="open"
>
  <div
    class="main-nav-container"
  >
    <div
      class="nav-group"
    >
      <div
        class="nav-menu-item"
        id="my-work-nav-item"
        target="_blank"
      >
        <div
          appearance="light"
          icon-name="sparkles"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-strong-md"
          >
            My work
          </span>
        </div>
        <div
          hidden=""
          slot="submenu"
        />
      </div>
      <div
        anchor="my-work-nav-item"
        placement="right"
      >
        My work
      </div>
      <div
        class="nav-menu-item"
        id="matters-nav-item"
        target="_blank"
      >
        <div
          appearance="light"
          icon-name="database"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-strong-md"
          >
            Databases
          </span>
        </div>
      </div>
      <div
        anchor="matters-nav-item"
        placement="right"
      >
        Databases
      </div>
      <div
        class="nav-menu-item"
        id="manage-nav-item"
        target="_blank"
      >
        <div
          appearance="light"
          icon-name="gear"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-strong-md"
          >
            Manage ShopAIAssist
          </span>
        </div>
        <div
          hidden=""
          slot="submenu"
        />
      </div>
      <div
        anchor="manage-nav-item"
        placement="right"
      >
        Manage ShopAIAssist
      </div>
    </div>
    <div
      class="nav-group"
    >
      <div
        class="nav-menu-item"
        id="whats-new-nav-item"
        target="_blank"
        url="https://www.example.com/en/release-notes.html"
      >
        <div
          appearance="light"
          icon-name="bell"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-strong-md"
          >
            What's new
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        anchor="whats-new-nav-item"
        placement="right"
      >
        What's new
      </div>
      <div
        class="nav-menu-item"
        id="support-nav-item"
        target="_blank"
      >
        <div
          appearance="light"
          icon-name="circle-question"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-strong-md"
          >
            Support
          </span>
        </div>
        <div
          hidden=""
          slot="submenu"
        />
      </div>
      <div
        anchor="support-nav-item"
        placement="right"
      >
        Support
      </div>
      <div
        class="nav-menu-item"
        id="legal-nav-item"
        target="_blank"
      >
        <div
          appearance="light"
          icon-name="shield"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-strong-md"
          >
            Legal information
          </span>
        </div>
        <div
          hidden=""
          slot="submenu"
        />
      </div>
      <div
        anchor="legal-nav-item"
        placement="right"
      >
        Legal information
      </div>
      <footer
        class="main-footer"
      >
        <span
          appearance="body-default-xs"
          class="copyright"
        >
          Â© 2024 SHOPAIASSISTReuters.
          <br />
           
          All rights reserved.
        </span>
      </footer>
    </div>
  </div>
</div>
`;
